parlai/core/torch_ranker_agent.py

```diff
@@ -28,7 +28,7 @@ def add_cmdline_args(argparser):
                  '(see TorchRankerAgent._build_candidates() for details).')
         agent.add_argument(
             '-ecands', '--eval-candidates', type=str,
-            choices=['batch', 'inline', 'fixed', 'vocab'],
+            choices=['batch', 'inline', 'fixed', 'vocab', 'custom'],
```

Comment:
I'm still confused w.r.t. all of this, and mostly wondering what's the logic here, and whether all this extra logic should be given a name and moved into TRA.

Reply:
the logic is in TRA for custom candidates

Reply:
Ahh, mostly I just don't understand what Starspace is trying to do w.r.t negative sampling, and whether it should be supported outside of starspace.